In [1]:
%matplotlib inline
%config InlineBackend.figure_format = 'retina'
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [3]:
df_train = pd.read_csv('data/train.csv')

In [4]:
df_train.head()

id  vendor_id      pickup_datetime     dropoff_datetime  \
0  id2875421          2  2016-03-14 17:24:55  2016-03-14 17:32:30   
1  id2377394          1  2016-06-12 00:43:35  2016-06-12 00:54:38   
2  id3858529          2  2016-01-19 11:35:24  2016-01-19 12:10:48   
3  id3504673          2  2016-04-06 19:32:31  2016-04-06 19:39:40   
4  id2181028          2  2016-03-26 13:30:55  2016-03-26 13:38:10   

   passenger_count  pickup_longitude  pickup_latitude  dropoff_longitude  \
0                1        -73.982155        40.767937         -73.964630   
1                1        -73.980415        40.738564         -73.999481   
2                1        -73.979027        40.763939         -74.005333   
3                1        -74.010040        40.719971         -74.012268   
4                1        -73.973053        40.793209         -73.972923   

   dropoff_latitude store_and_fwd_flag  trip_duration  
0         40.765602                  N            455  
1         40.731152                  N            663  
2         40.710087                  N           2124  
3         40.706718                  N            429  
4         40.782520                  N            435

In [5]:
df_train.describe(include='all')

id     vendor_id      pickup_datetime     dropoff_datetime  \
count     1458644  1.458644e+06              1458644              1458644   
unique    1458644           NaN              1380222              1380377   
top     id3368889           NaN  2016-01-12 18:48:44  2016-05-16 19:40:28   
freq            1           NaN                    5                    5   
mean          NaN  1.534950e+00                  NaN                  NaN   
std           NaN  4.987772e-01                  NaN                  NaN   
min           NaN  1.000000e+00                  NaN                  NaN   
25%           NaN  1.000000e+00                  NaN                  NaN   
50%           NaN  2.000000e+00                  NaN                  NaN   
75%           NaN  2.000000e+00                  NaN                  NaN   
max           NaN  2.000000e+00                  NaN                  NaN   

        passenger_count  pickup_longitude  pickup_latitude  dropoff_longitude  \
count      1.458644e+06      1.458644e+06     1.458644e+06       1.458644e+06   
unique              NaN               NaN              NaN                NaN   
top                 NaN               NaN              NaN                NaN   
freq                NaN               NaN              NaN                NaN   
mean       1.664530e+00     -7.397349e+01     4.075092e+01      -7.397342e+01   
std        1.314242e+00      7.090186e-02     3.288119e-02       7.064327e-02   
min        0.000000e+00     -1.219333e+02     3.435970e+01      -1.219333e+02   
25%        1.000000e+00     -7.399187e+01     4.073735e+01      -7.399133e+01   
50%        1.000000e+00     -7.398174e+01     4.075410e+01      -7.397975e+01   
75%        2.000000e+00     -7.396733e+01     4.076836e+01      -7.396301e+01   
max        9.000000e+00     -6.133553e+01     5.188108e+01      -6.133553e+01   

        dropoff_latitude store_and_fwd_flag  trip_duration  
count       1.458644e+06            1458644   1.458644e+06  
unique               NaN                  2            NaN  
top                  NaN                  N            NaN  
freq                 NaN            1450599            NaN  
mean        4.075180e+01                NaN   9.594923e+02  
std         3.589056e-02                NaN   5.237432e+03  
min         3.218114e+01                NaN   1.000000e+00  
25%         4.073588e+01                NaN   3.970000e+02  
50%         4.075452e+01                NaN   6.620000e+02  
75%         4.076981e+01                NaN   1.075000e+03  
max         4.392103e+01                NaN   3.526282e+06

In [6]:
df_train['store_and_fwd_flag'].value_counts()

N    1450599
Y       8045
Name: store_and_fwd_flag, dtype: int64

- vendor_id: OK (2 values)
- pickup_datetime: Split into (year, month, date, day of week, hour, minute (10 scale))
- dropoff_datetime: Split into (year, month, date, day of week, hour, minute (10 scale))
- passenger_count: (0-4-9 values)
- pickup_longitude: Normalize
- pickup_latitude: Normalize
- dropoff_longitude: Normalize
- dropoff_latitude: Normalize
- store_and_fwd_flag: Binarize

In [7]:
df_pickup_datetime = df_train['pickup_datetime'].apply(lambda x: x.replace('-',' ').replace(':',' ').split())
df_train['pickup_year'] = df_pickup_datetime.apply(lambda x: int(x[0]))
df_train['pickup_month'] = df_pickup_datetime.apply(lambda x: int(x[1]))
df_train['pickup_day'] = df_pickup_datetime.apply(lambda x: int(x[2]))
df_train['pickup_hour'] = df_pickup_datetime.apply(lambda x: int(x[3]))
df_train['pickup_minute'] = df_pickup_datetime.apply(lambda x: int(x[4]))
# df['second'] = df_pickup_datetime.apply(lambda x: x[5])

In [8]:
df_dropoff_datetime = df_train['dropoff_datetime'].apply(lambda x: x.replace('-',' ').replace(':',' ').split())
df_train['dropoff_year'] = df_dropoff_datetime.apply(lambda x: int(x[0]))
df_train['dropoff_month'] = df_dropoff_datetime.apply(lambda x: int(x[1]))
df_train['dropoff_day'] = df_dropoff_datetime.apply(lambda x: int(x[2]))
df_train['dropoff_hour'] = df_dropoff_datetime.apply(lambda x: int(x[3]))
df_train['dropoff_minute'] = df_dropoff_datetime.apply(lambda x: int(x[4]))
# df['second'] = df_pickup_datetime.apply(lambda x: x[5])

In [9]:
df_train['store_and_fwd_flag'] = df_train['store_and_fwd_flag'].apply(lambda x: 0 if x=="N" else 1 )

In [10]:
df_train.head()

id  vendor_id      pickup_datetime     dropoff_datetime  \
0  id2875421          2  2016-03-14 17:24:55  2016-03-14 17:32:30   
1  id2377394          1  2016-06-12 00:43:35  2016-06-12 00:54:38   
2  id3858529          2  2016-01-19 11:35:24  2016-01-19 12:10:48   
3  id3504673          2  2016-04-06 19:32:31  2016-04-06 19:39:40   
4  id2181028          2  2016-03-26 13:30:55  2016-03-26 13:38:10   

   passenger_count  pickup_longitude  pickup_latitude  dropoff_longitude  \
0                1        -73.982155        40.767937         -73.964630   
1                1        -73.980415        40.738564         -73.999481   
2                1        -73.979027        40.763939         -74.005333   
3                1        -74.010040        40.719971         -74.012268   
4                1        -73.973053        40.793209         -73.972923   

   dropoff_latitude  store_and_fwd_flag       ...        pickup_year  \
0         40.765602                   0       ...               2016   
1         40.731152                   0       ...               2016   
2         40.710087                   0       ...               2016   
3         40.706718                   0       ...               2016   
4         40.782520                   0       ...               2016   

   pickup_month  pickup_day  pickup_hour  pickup_minute  dropoff_year  \
0             3          14           17             24          2016   
1             6          12            0             43          2016   
2             1          19           11             35          2016   
3             4           6           19             32          2016   
4             3          26           13             30          2016   

   dropoff_month  dropoff_day  dropoff_hour  dropoff_minute  
0              3           14            17              32  
1              6           12             0              54  
2              1           19            12              10  
3              4            6            19              39  
4              3           26            13              38  

[5 rows x 21 columns]

In [11]:
df_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1458644 entries, 0 to 1458643
Data columns (total 21 columns):
id                    1458644 non-null object
vendor_id             1458644 non-null int64
pickup_datetime       1458644 non-null object
dropoff_datetime      1458644 non-null object
passenger_count       1458644 non-null int64
pickup_longitude      1458644 non-null float64
pickup_latitude       1458644 non-null float64
dropoff_longitude     1458644 non-null float64
dropoff_latitude      1458644 non-null float64
store_and_fwd_flag    1458644 non-null int64
trip_duration         1458644 non-null int64
pickup_year           1458644 non-null int64
pickup_month          1458644 non-null int64
pickup_day            1458644 non-null int64
pickup_hour           1458644 non-null int64
pickup_minute         1458644 non-null int64
dropoff_year          1458644 non-null int64
dropoff_month         1458644 non-null int64
dropoff_day           1458644 non-null int64
dropoff_hour         

In [12]:
df_ml = df_train.drop(['id', 'pickup_datetime', 'dropoff_datetime'], axis=1)

In [13]:
df_ml.head()

vendor_id  passenger_count  pickup_longitude  pickup_latitude  \
0          2                1        -73.982155        40.767937   
1          1                1        -73.980415        40.738564   
2          2                1        -73.979027        40.763939   
3          2                1        -74.010040        40.719971   
4          2                1        -73.973053        40.793209   

   dropoff_longitude  dropoff_latitude  store_and_fwd_flag  trip_duration  \
0         -73.964630         40.765602                   0            455   
1         -73.999481         40.731152                   0            663   
2         -74.005333         40.710087                   0           2124   
3         -74.012268         40.706718                   0            429   
4         -73.972923         40.782520                   0            435   

   pickup_year  pickup_month  pickup_day  pickup_hour  pickup_minute  \
0         2016             3          14           17             24   
1         2016             6          12            0             43   
2         2016             1          19           11             35   
3         2016             4           6           19             32   
4         2016             3          26           13             30   

   dropoff_year  dropoff_month  dropoff_day  dropoff_hour  dropoff_minute  
0          2016              3           14            17              32  
1          2016              6           12             0              54  
2          2016              1           19            12              10  
3          2016              4            6            19              39  
4          2016              3           26            13              38

In [14]:
from sklearn.preprocessing import StandardScaler

X, y = df_ml.drop('trip_duration', axis=1), df_ml['trip_duration']

normalizer = StandardScaler()
X = normalizer.fit_transform(X)

In [15]:
pd.DataFrame(X).head()

0         1         2         3         4         5         6    7   \
0  0.932380 -0.505637 -0.122261  0.517494  0.124369  0.384575 -0.074471  0.0   
1 -1.072524 -0.505637 -0.097727 -0.375819 -0.368970 -0.575303 -0.074471  0.0   
2  0.932380 -0.505637 -0.078143  0.395910 -0.451805 -1.162220 -0.074471  0.0   
3  0.932380 -0.505637 -0.515558 -0.941274 -0.549976 -1.256071 -0.074471  0.0   
4  0.932380 -0.505637  0.006112  1.286091  0.006974  0.855957 -0.074471  0.0   

         8         9         10        11   12        13        14        15  \
0 -0.307440 -0.172813  0.530263 -0.322670  0.0 -0.307621 -0.172867  0.524563   
1  1.477173 -0.402616 -2.126116  0.774030  0.0  1.476800 -0.402670 -2.097405   
2 -1.497182  0.401692 -0.407283  0.312262  0.0 -1.497236  0.401639 -0.246604   
3  0.287431 -1.092023  0.842778  0.139099  0.0  0.287186 -1.092078  0.833030   
4 -0.307440  1.206001 -0.094768  0.023657  0.0 -0.307621  1.205949 -0.092371   

         16  
0  0.136139  
1  1.399039  
2 -1.126761  
3  0.537971  
4  0.480566

In [16]:
from sklearn.model_selection import cross_val_score
from sklearn.linear_model import LinearRegression
from sklearn.linear_model import Ridge
from sklearn.linear_model import Lasso
from sklearn.linear_model import ElasticNet
from sklearn.linear_model import SGDRegressor
from sklearn.svm import SVR
from sklearn.kernel_ridge import KernelRidge
from sklearn.neighbors import KNeighborsRegressor
from sklearn.gaussian_process import GaussianProcessRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.neural_network import MLPRegressor
from xgboost import XGBRegressor
import time

/home/alulab/.conda/envs/kaggle/lib/python3.6/site-packages/sklearn/cross_validation.py:41: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. Also note that the interface of the new CV iterators are different from that of this module. This module will be removed in 0.20.
  "This module will be removed in 0.20.", DeprecationWarning)


In [23]:
import locale
locale.setlocale(locale.LC_ALL, 'en_US.utf8')

'en_US.utf8'

In [38]:
number = 1861616151
type(locale.format("%d", number, grouping=True))

str

In [39]:
def run_model(clf, X, y):
    start = int(time.time() * 1000)
    scores = cross_val_score(clf, X, y, cv=5, scoring='neg_mean_squared_error')
    end = int(time.time() * 1000)
    print("%s \t\t MSE: %s (+/- %s) Time: %s ms" % \
          (str(clf.__class__).split('.')[-1].replace('>','').replace("'",''), \
          locale.format("%0.2f", scores.mean(), grouping=True),
          locale.format("%0.2f", scores.std() * 2, grouping=True),
          locale.format("%d", float(end - start), grouping=True)))
          #end - start))

In [40]:
def run_models(X, y):
    run_model(LinearRegression(), X, y)
    run_model(Ridge(), X, y)
#     run_model(Lasso(), X, y)
#     run_model(ElasticNet(), X, y)
#     run_model(SVR(verbose=True), X, y)
    run_model(SGDRegressor(max_iter=10), X, y)
#     run_model(KernelRidge(), X, y)
#     run_model(KNeighborsRegressor(), X, y)
    run_model(XGBRegressor(silent=False), X, y)
#     run_model(GaussianProcessRegressor(), X, y)
    run_model(RandomForestRegressor(verbose=1), X, y)
#     run_model(GradientBoostingRegressor(verbose=1), X, y)    
#     run_model(MLPRegressor(verbose=True), X, y)

In [41]:
run_models(X, y)

LinearRegression 		 MSE: -1,105,289.53 (+/- 256,791.21) Time: 5,952 ms
Ridge 		 MSE: -10,396,452.23 (+/- 21,157,800.77) Time: 2,336 ms
SGDRegressor 		 MSE: -27,488,561.64 (+/- 43,994,785.47) Time: 14,044 ms
XGBRegressor 		 MSE: -23,028,788.68 (+/- 36,457,345.71) Time: 163,129 ms


[Parallel(n_jobs=1)]: Done  10 out of  10 | elapsed:  2.8min finished
[Parallel(n_jobs=1)]: Done  10 out of  10 | elapsed:    1.7s finished
[Parallel(n_jobs=1)]: Done  10 out of  10 | elapsed:  2.7min finished
[Parallel(n_jobs=1)]: Done  10 out of  10 | elapsed:    1.7s finished
[Parallel(n_jobs=1)]: Done  10 out of  10 | elapsed:  2.7min finished
[Parallel(n_jobs=1)]: Done  10 out of  10 | elapsed:    1.8s finished
[Parallel(n_jobs=1)]: Done  10 out of  10 | elapsed:  2.7min finished
[Parallel(n_jobs=1)]: Done  10 out of  10 | elapsed:    1.6s finished
[Parallel(n_jobs=1)]: Done  10 out of  10 | elapsed:  2.7min finished


RandomForestRegressor 		 MSE: -31,681,050.95 (+/- 41,293,820.50) Time: 826,227 ms


[Parallel(n_jobs=1)]: Done  10 out of  10 | elapsed:    1.8s finished


- LinearRegression 		 MSE: -1105286.49 (+/- 256792.45) Time: 7119 ms
- Ridge 		 MSE: -10396438.11 (+/- 21157617.24) Time: 3755 ms
- Lasso 		 MSE: -27346063.43 (+/- 43753148.12) Time: 296502 ms
- ElasticNet 		 MSE: -27379607.56 (+/- 43784875.68) Time: 9865 ms
- SGDRegressor 		 MSE: -27496244.81 (+/- 43989131.50) Time: 10100 ms
- XGBRegressor 		 MSE: -23028788.68 (+/- 36457345.71) Time: 431149 ms

## つぎ

In [22]:
from datetime import datetime
to_weekday = ['Lunes', 'Martes', 'Miercoles', 'Jueves', 'Viernes', 'Sabado', 'Domingo']
def get_weekday(fecha):
    datetime_obj = datetime.strptime(fecha.split()[0], "%Y-%m-%d")
#     return to_weekday[datetime_obj.weekday()]
    return datetime_obj.weekday()

In [23]:
df_train['pickup_weekday'] = df_train['pickup_datetime'].apply(get_weekday)
df_train['dropoff_weekday'] = df_train['dropoff_datetime'].apply(get_weekday)

In [24]:
df_train.head()

id  vendor_id      pickup_datetime     dropoff_datetime  \
0  id2875421          2  2016-03-14 17:24:55  2016-03-14 17:32:30   
1  id2377394          1  2016-06-12 00:43:35  2016-06-12 00:54:38   
2  id3858529          2  2016-01-19 11:35:24  2016-01-19 12:10:48   
3  id3504673          2  2016-04-06 19:32:31  2016-04-06 19:39:40   
4  id2181028          2  2016-03-26 13:30:55  2016-03-26 13:38:10   

   passenger_count  pickup_longitude  pickup_latitude  dropoff_longitude  \
0                1        -73.982155        40.767937         -73.964630   
1                1        -73.980415        40.738564         -73.999481   
2                1        -73.979027        40.763939         -74.005333   
3                1        -74.010040        40.719971         -74.012268   
4                1        -73.973053        40.793209         -73.972923   

   dropoff_latitude  store_and_fwd_flag       ...         pickup_day  \
0         40.765602                   0       ...                 14   
1         40.731152                   0       ...                 12   
2         40.710087                   0       ...                 19   
3         40.706718                   0       ...                  6   
4         40.782520                   0       ...                 26   

   pickup_hour  pickup_minute  dropoff_year  dropoff_month  dropoff_day  \
0           17             24          2016              3           14   
1            0             43          2016              6           12   
2           11             35          2016              1           19   
3           19             32          2016              4            6   
4           13             30          2016              3           26   

   dropoff_hour  dropoff_minute  pickup_weekday  dropoff_weekday  
0            17              32               0                0  
1             0              54               6                6  
2            12              10               1                1  
3            19              39               2                2  
4            13              38               5                5  

[5 rows x 23 columns]

In [25]:
from math import radians, cos, sin, asin, sqrt
def haversine(lon1, lat1, lon2, lat2):
    lon1, lat1, lon2, lat2 = map(radians, [lon1, lat1, lon2, lat2])
    dlon = lon2 - lon1 
    dlat = lat2 - lat1 
    a = sin(dlat/2)**2 + cos(lat1) * cos(lat2) * sin(dlon/2)**2
    c = 2 * asin(sqrt(a)) 
    km = 6367 * c
    return km

In [33]:
df_train['distance'] = df_train.apply(haversine(row['pickup_longitude'], row['pickup_latitude'],
                                                row['dropoff_longitude'], row['dropoff_latitude']), axis=1)

In [34]:
df_train.head()

id  vendor_id      pickup_datetime     dropoff_datetime  \
0  id2875421          2  2016-03-14 17:24:55  2016-03-14 17:32:30   
1  id2377394          1  2016-06-12 00:43:35  2016-06-12 00:54:38   
2  id3858529          2  2016-01-19 11:35:24  2016-01-19 12:10:48   
3  id3504673          2  2016-04-06 19:32:31  2016-04-06 19:39:40   
4  id2181028          2  2016-03-26 13:30:55  2016-03-26 13:38:10   

   passenger_count  pickup_longitude  pickup_latitude  dropoff_longitude  \
0                1        -73.982155        40.767937         -73.964630   
1                1        -73.980415        40.738564         -73.999481   
2                1        -73.979027        40.763939         -74.005333   
3                1        -74.010040        40.719971         -74.012268   
4                1        -73.973053        40.793209         -73.972923   

   dropoff_latitude  store_and_fwd_flag    ...     pickup_hour  pickup_minute  \
0         40.765602                   0    ...              17             24   
1         40.731152                   0    ...               0             43   
2         40.710087                   0    ...              11             35   
3         40.706718                   0    ...              19             32   
4         40.782520                   0    ...              13             30   

   dropoff_year  dropoff_month  dropoff_day  dropoff_hour  dropoff_minute  \
0          2016              3           14            17              32   
1          2016              6           12             0              54   
2          2016              1           19            12              10   
3          2016              4            6            19              39   
4          2016              3           26            13              38   

   pickup_weekday  dropoff_weekday  distance  
0               0                0  1.497580  
1               6                6  1.804374  
2               1                1  6.381090  
3               2                2  1.484566  
4               5                5  1.187842  

[5 rows x 24 columns]

In [35]:
df_ml = df_train.drop(['id', 'pickup_datetime', 'dropoff_datetime'], axis=1)

In [36]:
from sklearn.preprocessing import StandardScaler

X, y = df_ml.drop('trip_duration', axis=1), df_ml['trip_duration']

normalizer = StandardScaler()
X = normalizer.fit_transform(X)

In [37]:
X.shape

(1458644, 20)

In [39]:
run_models(X, y)

LinearRegression 		 MSE: -1101410.57 (+/- 262297.90) Time: 7808 ms
Ridge 		 MSE: -10403256.54 (+/- 21333332.40) Time: 5614 ms
SGDRegressor 		 MSE: -85921259.81 (+/- 267145368.78) Time: 20297 ms
XGBRegressor 		 MSE: -20811449.24 (+/- 38746219.06) Time: 475262 ms


### Test

In [40]:
df_test = pd.read_csv('test.csv')

In [41]:
df_test.head()

id  vendor_id      pickup_datetime  passenger_count  \
0  id3004672          1  2016-06-30 23:59:58                1   
1  id3505355          1  2016-06-30 23:59:53                1   
2  id1217141          1  2016-06-30 23:59:47                1   
3  id2150126          2  2016-06-30 23:59:41                1   
4  id1598245          1  2016-06-30 23:59:33                1   

   pickup_longitude  pickup_latitude  dropoff_longitude  dropoff_latitude  \
0        -73.988129        40.732029         -73.990173         40.756680   
1        -73.964203        40.679993         -73.959808         40.655403   
2        -73.997437        40.737583         -73.986160         40.729523   
3        -73.956070        40.771900         -73.986427         40.730469   
4        -73.970215        40.761475         -73.961510         40.755890   

  store_and_fwd_flag  
0                  N  
1                  N  
2                  N  
3                  N  
4                  N

In [48]:
def preprocess(df):
    df_pickup_datetime = df['pickup_datetime'].apply(lambda x: x.replace('-',' ').replace(':',' ').split())
    df['pickup_year'] = df_pickup_datetime.apply(lambda x: int(x[0]))
    df['pickup_month'] = df_pickup_datetime.apply(lambda x: int(x[1]))
    df['pickup_day'] = df_pickup_datetime.apply(lambda x: int(x[2]))
    df['pickup_hour'] = df_pickup_datetime.apply(lambda x: int(x[3]))
    df['pickup_minute'] = df_pickup_datetime.apply(lambda x: int(x[4]))
    df['store_and_fwd_flag'] = df['store_and_fwd_flag'].apply(lambda x: 0 if x=="N" else 1)
    df['pickup_weekday'] = df['pickup_datetime'].apply(get_weekday)
    df['distance'] = df.apply(lambda row: haversine(row['pickup_longitude'], row['pickup_latitude'],
                                                    row['dropoff_longitude'], row['dropoff_latitude']), axis=1)
    df = df.drop(['id', 'pickup_datetime'], axis=1)
    return df

In [49]:
df_train = pd.read_csv('train.csv')
df_train = preprocess(df_train)

In [50]:
df_train.head()

vendor_id     dropoff_datetime  passenger_count  pickup_longitude  \
0          2  2016-03-14 17:32:30                1        -73.982155   
1          1  2016-06-12 00:54:38                1        -73.980415   
2          2  2016-01-19 12:10:48                1        -73.979027   
3          2  2016-04-06 19:39:40                1        -74.010040   
4          2  2016-03-26 13:38:10                1        -73.973053   

   pickup_latitude  dropoff_longitude  dropoff_latitude  store_and_fwd_flag  \
0        40.767937         -73.964630         40.765602                   0   
1        40.738564         -73.999481         40.731152                   0   
2        40.763939         -74.005333         40.710087                   0   
3        40.719971         -74.012268         40.706718                   0   
4        40.793209         -73.972923         40.782520                   0   

   trip_duration  pickup_year  pickup_month  pickup_day  pickup_hour  \
0            455         2016             3          14           17   
1            663         2016             6          12            0   
2           2124         2016             1          19           11   
3            429         2016             4           6           19   
4            435         2016             3          26           13   

   pickup_minute  pickup_weekday  distance  
0             24               0  1.497580  
1             43               6  1.804374  
2             35               1  6.381090  
3             32               2  1.484566  
4             30               5  1.187842

In [52]:
from sklearn.preprocessing import StandardScaler
X, y = df_train.drop(['trip_duration', 'dropoff_datetime'], axis=1), df_train['trip_duration']

normalizer = StandardScaler()
X = normalizer.fit_transform(X)

In [66]:
from sklearn.metrics import mean_squared_error

In [69]:
reg = LinearRegression()
reg.fit(X, y)
y_pred = reg.predict(X)
mean_squared_error(y, y_pred)

27168912.53498758

In [70]:
xgb = XGBRegressor()
xgb.fit(X, y)
y_pred = xgb.predict(X)
mean_squared_error(y, y_pred)

19047522.590025336

In [71]:
df_test = pd.read_csv('test.csv')
df_test = preprocess(df_test)

In [72]:
df_test.head()

vendor_id  passenger_count  pickup_longitude  pickup_latitude  \
0          1                1        -73.988129        40.732029   
1          1                1        -73.964203        40.679993   
2          1                1        -73.997437        40.737583   
3          2                1        -73.956070        40.771900   
4          1                1        -73.970215        40.761475   

   dropoff_longitude  dropoff_latitude  store_and_fwd_flag  pickup_year  \
0         -73.990173         40.756680                   0         2016   
1         -73.959808         40.655403                   0         2016   
2         -73.986160         40.729523                   0         2016   
3         -73.986427         40.730469                   0         2016   
4         -73.961510         40.755890                   0         2016   

   pickup_month  pickup_day  pickup_hour  pickup_minute  pickup_weekday  \
0             6          30           23             59               3   
1             6          30           23             59               3   
2             6          30           23             59               3   
3             6          30           23             59               3   
4             6          30           23             59               3   

   distance  
0  2.744701  
1  2.757507  
2  1.305335  
3  5.265780  
4  0.960239

In [73]:
X_test = normalizer.transform(df_test)

In [74]:
y_pred_reg = reg.predict(X_test)

In [75]:
y_pred_xgb = xgb.predict(X_test)

In [100]:
y_pred_reg = np.maximum(y_pred_reg, 1)
y_pred_xgb = np.maximum(y_pred_xgb, 1)

In [76]:
df_id = pd.read_csv('test.csv')

In [103]:
pd.DataFrame(np.hstack((np.reshape(df_id['id'].values, (-1,1)), np.reshape(y_pred_reg, (-1,1)))), 
             columns=['id', 'trip_duration']).to_csv('reg_submission.csv', index=False)

In [104]:
pd.DataFrame(np.hstack((np.reshape(df_id['id'].values, (-1,1)), np.reshape(y_pred_xgb, (-1,1)))), 
             columns=['id', 'trip_duration']).to_csv('xgb_submission.csv', index=False)

In [101]:
y_pred_reg[y_pred_reg<0]

array([], dtype=float64)

In [102]:
y_pred_xgb[y_pred_xgb<0]

array([], dtype=float32)